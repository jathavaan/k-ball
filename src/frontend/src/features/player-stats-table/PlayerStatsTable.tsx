import { TableBody, TableCell, TableRow, Paper, TableContainer } from "@mui/material";
import { PlayerStatsProps, PlayerStatsTableProps } from "./playerStatsTable.types";
import { StyledPlayerStatsTable, StyledTableHead } from "./playerStatsTable.style";

export const PlayerStatsTable = (props: PlayerStatsTableProps) => {
  const { playerStatsTable } = props; 

  return (
    <TableContainer component={Paper}>
      <StyledPlayerStatsTable>
        <StyledTableHead>
          <TableRow>
            <TableCell>Season</TableCell>
            <TableCell align="right">Goals</TableCell>
            <TableCell align="right">Assists</TableCell>
            <TableCell align="right">Appearances</TableCell>
            <TableCell align="right">Yellow Cards</TableCell>
            <TableCell align="right">Red Cards</TableCell>
          </TableRow>
        </StyledTableHead>
        <TableBody>
          {playerStatsTable.map((row: PlayerStatsProps) => (
            <TableRow key={row.season}>
              <TableCell component="th" scope="row">
                {row.season}
              </TableCell>
              <TableCell align="right">{row.goals}</TableCell>
              <TableCell align="right">{row.assists}</TableCell>
              <TableCell align="right">{row.appearances}</TableCell>
              <TableCell align="right">{row.yellowCards}</TableCell>
              <TableCell align="right">{row.redCards}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </StyledPlayerStatsTable>
    </TableContainer>
  );
};
