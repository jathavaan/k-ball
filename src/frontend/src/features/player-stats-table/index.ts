export * from "./playerStatsTable.slice.ts";
export * from "./PlayerStatsTable.tsx";
export * from "./playerStatsTable.hooks.ts";
