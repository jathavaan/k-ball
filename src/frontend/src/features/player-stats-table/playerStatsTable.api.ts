import { PlayerStatsTableProps } from "@features/player-stats-table/playerStatsTable.types";

export const getPlayerStatsTable = ({
  queryKey,
}: {
  queryKey: [string, number];
}): Promise<PlayerStatsTableProps> => {
  const [, playerId] = queryKey;

  // Helper function to generate random stat values within defined ranges
  const generateRandomStats = (season: number) => ({
    playerId: playerId,
    season,
    goals: Math.floor(Math.random() * 15), // 0-14 goals
    assists: Math.floor(Math.random() * 10), // 0-9 assists
    appearances: Math.floor(Math.random() * 30) + 1, // 1-30 appearances
    yellowCards: Math.floor(Math.random() * 5), // 0-4 yellow cards
    redCards: Math.floor(Math.random() * 2), // 0-1 red cards
  });

  return new Promise((resolve) => {
    setTimeout(() => {
      const seasons = [2023, 2022, 2020, 2019, 2018];
      const playerStats = seasons.map(generateRandomStats);

      resolve({ playerStatsTable: playerStats });
    }, 1);
  });
};
