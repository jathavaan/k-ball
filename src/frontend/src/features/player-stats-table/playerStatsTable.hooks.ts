import { AppDispatch } from "@/store.ts";
import { useDispatch } from "react-redux";
import { setPlayerStatsTable } from "@features/player-stats-table/playerStatsTable.slice";
import { useEffect } from "react";
import { usePlayerStatsTable } from "@features/player-stats-table/playerStatsTable.query";

export const usePlayerStatsTableData = (playerId: number) => {
  const dispatch = useDispatch<AppDispatch>();
  const {
    data: playerStatsTable,
    isLoading,
    isError,
  } = usePlayerStatsTable(playerId);

  useEffect(() => {
    if (playerStatsTable) {
      dispatch(setPlayerStatsTable(playerStatsTable));
    }
  }, [dispatch, playerStatsTable]);

  return { isLoading, isError };
};
