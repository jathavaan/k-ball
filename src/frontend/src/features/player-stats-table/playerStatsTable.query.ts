import { useQuery } from "@tanstack/react-query";
import { getPlayerStatsTable } from "./playerStatsTable.api";

export const usePlayerStatsTable = (playerId: number) =>
  useQuery({
    queryKey: ["playerStatsTables", playerId],
    queryFn: getPlayerStatsTable,
  });
