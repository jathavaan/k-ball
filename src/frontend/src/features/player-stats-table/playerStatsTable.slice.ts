import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import {
  PlayerStatsProps,
  PlayerStatsTableState,
} from "./playerStatsTable.types";

const initialState: PlayerStatsTableState = {
  playerStatsTable: [],
};

const playerStatsTableSlice = createSlice({
  name: "playerStatsTables",
  initialState,
  reducers: {
    setPlayerStatsTable: (
      state,
      action: PayloadAction<{ playerStatsTable: PlayerStatsProps[] }>,
    ) => {
      state.playerStatsTable = action.payload.playerStatsTable;
    },
  },
});

export const { setPlayerStatsTable } = playerStatsTableSlice.actions;
export const playerStatsTableReducer = playerStatsTableSlice.reducer;
