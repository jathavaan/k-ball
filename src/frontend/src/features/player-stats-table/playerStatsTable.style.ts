import {
  Table,
  TableHead,
  TableRow,
  TableCell,
  TableBody,
  styled,
} from "@mui/material";

export const StyledPlayerStatsTable = styled(Table)(({ theme }) => ({
  backgroundColor: theme.palette.primary.light,
  color: theme.palette.primary.contrastText,
  width: "100%",
  borderCollapse: "collapse",
  borderRadius: "8px",
  overflow: "hidden",
  [theme.breakpoints.down("sm")]: {
    fontSize: "0.75rem",
  },
}));
export const StyledTableHead = styled(TableHead)(({ theme }) => ({
  backgroundColor: theme.palette.secondary.main,
  color: theme.palette.primary.contrastText,
  "& .MuiTableCell-head": {
    fontWeight: "bold",
    fontSize: "1rem",
    padding: "12px",
    textAlign: "center",
    [theme.breakpoints.down("sm")]: {
      fontSize: "0.85rem",
      padding: "8px",
    },
  },
}));

export const StyledTableCell = styled(TableCell)(({ theme }) => ({
  color: theme.palette.primary.contrastText,
  fontSize: "0.875rem",
  padding: "10px",
  borderBottom: `1px solid ${theme.palette.divider}`,
  textAlign: "center",
  [theme.breakpoints.down("sm")]: {
    fontSize: "0.75rem",
    padding: "6px",
  },
}));

export const StyledTableBody = styled(TableBody)(({ theme }) => ({
  "& .MuiTableRow-root": {
    "&:nth-of-type(odd)": {
      backgroundColor: theme.palette.action.hover,
    },
    "&:hover": {
      backgroundColor: theme.palette.secondary.light,
    },
    [theme.breakpoints.down("sm")]: {
      fontSize: "0.75rem",
    },
  },
}));

export const StyledTableRow = styled(TableRow)(() => ({
  "&:last-child td, &:last-child th": {
    border: 0,
  },
}));
