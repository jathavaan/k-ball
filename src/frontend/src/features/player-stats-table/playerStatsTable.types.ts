import { TableProps } from "@mui/material";

export interface PlayerStatsProps {
  season: number;
  goals: number;
  assists: number;
  appearances: number;
  yellowCards: number;
  redCards: number;
}

export interface PlayerStatsTableProps extends TableProps {
  playerStatsTable: PlayerStatsProps[];
}

export interface PlayerStatsTableState {
  playerStatsTable: PlayerStatsProps[] | undefined;
}
