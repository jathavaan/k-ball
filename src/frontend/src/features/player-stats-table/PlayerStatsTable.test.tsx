import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { beforeEach, describe, expect, it } from "vitest";
import { PlayerStatsTable } from "@features/player-stats-table/PlayerStatsTable";
import { PlayerStatsProps } from "@features/player-stats-table/playerStatsTable.types";

describe("PlayerStatsTable", () => {
  const mockStats: PlayerStatsProps[] = [
    {
      playerId: 2898,
      season: 2021,
      goals: 3,
      assists: 2,
      appearances: 10,
      yellowCards: 1,
      redCards: 0,
    },
    {
      playerId: 2898,
      season: 2020,
      goals: 5,
      assists: 4,
      appearances: 15,
      yellowCards: 2,
      redCards: 0,
    },
  ];

  beforeEach(() => {
    render(<PlayerStatsTable playerStatsTable={mockStats} />);
  });

  it("should render the correct number of rows", () => {
    const rows = screen.getAllByRole("row");
    expect(rows.length).toBe(3);
  });

  it("should render the correct stats for the 2021 season", () => {
    const season2021 = screen.getByText("2021");
    expect(season2021).toBeInTheDocument();

    const goals2021 = screen.getByText("3");
    expect(goals2021).toBeInTheDocument();

    const assists2021 = screen.getAllByText("2")[0];
    expect(assists2021).toBeInTheDocument();
  });

  it("should render the correct stats for the 2020 season", () => {
    const season2020 = screen.getByText("2020");
    expect(season2020).toBeInTheDocument();

    const goals2020 = screen.getByText("5");
    expect(goals2020).toBeInTheDocument();

    const yellowCards2020 = screen.getAllByText("2")[1];
    expect(yellowCards2020).toBeInTheDocument();
  });

  it("should match the snapshot", () => {
    const { asFragment } = render(
      <PlayerStatsTable playerStatsTable={mockStats} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
