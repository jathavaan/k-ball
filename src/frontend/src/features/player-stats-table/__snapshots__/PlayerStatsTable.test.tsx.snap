// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayerStatsTable > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-16dlrry-MuiPaper-root-MuiTableContainer-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <table
      class="MuiTable-root css-ki6gk2-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-pv7sig-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-fe9koo-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1bwxs5v-MuiTableCell-root"
            scope="col"
          >
            Season
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-19yrrd1-MuiTableCell-root"
            scope="col"
          >
            Goals
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-19yrrd1-MuiTableCell-root"
            scope="col"
          >
            Assists
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-19yrrd1-MuiTableCell-root"
            scope="col"
          >
            Appearances
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-19yrrd1-MuiTableCell-root"
            scope="col"
          >
            Yellow Cards
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-19yrrd1-MuiTableCell-root"
            scope="col"
          >
            Red Cards
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-pae2gc-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-fe9koo-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1igmjpc-MuiTableCell-root"
            scope="row"
          >
            2021
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-118ta18-MuiTableCell-root"
          >
            3
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-118ta18-MuiTableCell-root"
          >
            2
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-118ta18-MuiTableCell-root"
          >
            10
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-118ta18-MuiTableCell-root"
          >
            1
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-118ta18-MuiTableCell-root"
          >
            0
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-fe9koo-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1igmjpc-MuiTableCell-root"
            scope="row"
          >
            2020
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-118ta18-MuiTableCell-root"
          >
            5
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-118ta18-MuiTableCell-root"
          >
            4
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-118ta18-MuiTableCell-root"
          >
            15
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-118ta18-MuiTableCell-root"
          >
            2
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-118ta18-MuiTableCell-root"
          >
            0
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
